infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -30,96 +39,206 @@ private StateTransfer() {
+                    segment.getStart(), segment.getEnd());
             return;
         }
 
-        int batchSize = runtime.getParameters().getBulkReadSize();
+        int chunkSize = runtime.getParameters().getBulkReadSize();
+        final AtomicInteger overwriteRetries = new AtomicInteger();
 
-        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
-        // Send the trimMark to the new/healing nodes.
-        // If this times out or fails, the Action performing the stateTransfer fails and retries.
-
-        endpoints.stream()
-                .map(endpoint -> {
-                    // TrimMark is the first address present on the log unit server.
-                    // Perform the prefix trim on the preceding address = (trimMark - 1).
-                    // Since the LU will reject trim decisions made from older epochs, we
-                    // need to adjust the new trim mark to have the new layout's epoch.
-                    Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
-                    return runtime.getLayoutView().getRuntimeLayout(layout)
-                            .getLogUnitClient(endpoint)
-                            .prefixTrim(prefixToken);
-                })
-                .forEach(CFUtils::getUninterruptibly);
-
-        if (trimMark > segment.getEnd()) {
-            log.info("stateTransfer: Nothing to transfer, trimMark {} greater than end of segment {}",
-                    trimMark, segment.getEnd());
-            return;
-        }
-
-        // State transfer should start from segment start address or trim mark whichever is lower.
-        long segmentStart = Math.max(trimMark, segment.getStart());
+        IRetry.build(ExponentialBackoffRetry.class, RetryExhaustedException.class, () -> {
 
-        for (long chunkStart = segmentStart; chunkStart < segment.getEnd()
-                ; chunkStart = chunkStart + batchSize) {
-            long chunkEnd = Math.min((chunkStart + batchSize - 1), segment.getEnd() - 1);
+            try {
 
-            long ts1 = System.currentTimeMillis();
+                long trimMark = setTrimOnNewLogUnit(layout, runtime, endpoint);
 
-            Map<Long, ILogData> dataMap = runtime.getAddressSpaceView()
-                    .fetchAll(ContiguousSet.create(Range.closed(chunkStart, chunkEnd),
-                            DiscreteDomain.longs()), true);
+                if (trimMark > segment.getEnd()) {
+                    log.info("stateTransfer: Nothing to transfer, trimMark {}"
+                                    + "greater than end of segment {}",
+                            trimMark, segment.getEnd());
+                    return true;
+                }
 
-            long ts2 = System.currentTimeMillis();
+                // State transfer should start from segment start address or trim mark
+                // whichever is lower.
+                final long segmentStart = Math.max(trimMark, segment.getStart());
```

Comment:
couldn't the trim mark be out of this segment? in which case it does not correspond to the given segment end?

Reply:
In that case, there is nothing to transfer.